pkg/roachprod/vm/gce/gcloud.go

```diff
@@ -473,6 +481,225 @@ func (p *Provider) GetVMSpecs(
+
+	mts.Lock()
+	defer mts.Unlock()
+
+	if mts.cache == nil {
+		mts.cache = make(map[string]*MachineTypeSpecs)
+	}
+
+	// We require an "{instance-type}-{project}-{zone}" combination to describe
+	// the instance type (as required by GCP), but store the cached results
+	// under two different keys:
+	// - a narrow one "{instance-type}-{project}-{zone}"
+	// - a wide one "{instance-type}-{project}"
+	// The justification behind this is that instance types are homogeneous
+	// across all zones, so we can increase cache hits when we don't actually
+	// care about the zone, but we can also request a narrow cache key when
+	// we want to ensure that this instance type is available in a specific zone
+	buildCacheKeys := func(machineType, project, zone string) map[CacheKeyLookup]string {
+		return map[CacheKeyLookup]string{
+			WideCacheKey:   machineType + "-" + project,
+			NarrowCacheKey: machineType + "-" + project + "-" + zone,
+		}
+	}
+	cacheKeys := buildCacheKeys(machineType, project, zone)
+
+	if cachedSpecs, ok := mts.cache[cacheKeys[ckl]]; ok {
+		return cachedSpecs, nil
+	}
+
+	// Query GCP to get machine type specs:
+	// Running:
+	// 		gcloud compute machine-types describe t2a-standard-4 \
+	//  		--project cockroach-ephemeral --zone us-east1-b \
+	// 			--format json
+	//
+	// Note: GCP requires a zone to describe a machine type, though during
+	// cluster creation, roachprod's takes a `--gce-machine-type` as input and
+	// guesses the CPU architecture from it. Only then, the zone is inferred
+	// from an hardcoded CPU architecture/zone mapping.
+	// As we're want to know the architecture from the machine type before
+	// knowing the zone, we try the default zone for both amd64 and arm64
+	// architectures. These calls are parallelized.
+	//
+	// In cases where we know the zone already (e.g. `roachprod list`), we use
+	// the supplied zone to query, making sure that we won't hit a zone in which
+	// the instance type is not available.
+
+	var zonesIteration []string
+	if zone != "" {
+		zonesIteration = []string{zone}
+	} else {
+		zonesIteration = []string{
+			DefaultZones(string(vm.ArchAMD64))[0],
+			DefaultZones(string(vm.ArchARM64))[0],
+		}
+	}
+
+	// gcloudQueryResult is the struct we use to pass around the query results.
+	// As we might iterate over multiple zones, we parallelize the calls and sink
+	// the results back via a channel.
+	type gcloudQueryResult struct {
+		specs *MachineTypeSpecs
+		err   error
+	}
+	resChan := make(chan gcloudQueryResult)
+	wg := sync.WaitGroup{}
+
+	for _, zone := range zonesIteration {
+		wg.Add(1)
+		go func(resChan chan<- gcloudQueryResult, machineType, project, zone string) {
+			args := []string{
+				"compute",
+				"machine-types",
+				"describe",
+				machineType,
+				"--project", project,
+				"--zone", zone,
+				"--format", "json",
+			}
+			specs := &MachineTypeSpecs{}
+			if err := runJSONCommand(args, &specs); err != nil {
+				if strings.Contains(err.Error(), "was not found") {
```

Comment:
Yikes. Maybe `not found` is more likely to stick around in future release? (Verb conjugation is kinda arbitrary.) Maybe there is specific error code instead?

Reply:
No specific error code, I checked.

Switching to just "not found".